chain/index/chain-indexing-overview-for-rpc-providers.MD

```diff
@@ -0,0 +1,202 @@
+# ChainIndexer Documentation for RPC Providers
+  
+## Introduction
+
+This document is for RPC providers and node operators who serve RPC requests walking through the configuration changes, migration flow and operations/maintenance work needed to enable, backfill and maintain the `ChainIndexer`.
+
+**Note: If you are a Storage Provider or node operator who does not serve RPC requests, you can skip this document as the `ChainIndexer` is already disabled by default**. 
+ 
+
+## ChainIndexer Config
+### Enablement
+
+The following must be enabled on an RPC node before starting as they are disabled by default:
+
+```toml
+[ChainIndexer]
+# Enable the ChainIndexer. 
+  EnableIndexer = true 
+
+[Fevm]
+# Enable the ETH RPC APIs.
+  EnableEthRPC = true
+
+[Events]
+# Enable the Actor Events APIs.
+  EnableActorEventsAPI = true
+```
+
+You can learn more about these configuration options and other configuration options available for the `ChainIndexer` [here](https://github.com/filecoin-project/lotus/blob/master/documentation/en/default-lotus-config.toml).
+
+
+### Garbage Collection
+
+The `ChainIndexer` includes a garbage collection (GC) mechanism to manage the amount of historical data retained. By default, GC is disabled to preserve all indexed data.
+
+To configure GC, use the `GCRetentionEpochs` parameter in the `ChainIndexer` section of your config.
+
+The ChainIndexer [periodically runs](https://github.com/filecoin-project/lotus/blob/master/chain/index/gc.go#L15) GC if `GCRetentionEpochs` is > 0 and removes indexed data for epochs older than `(current_head_height - GCRetentionEpochs)`.
+
+```toml
+[ChainIndexer]
+  GCRetentionEpochs = X  # Replace X with your desired value
+```
+
+- Setting `GCRetentionEpochs` to 0 (**default**) disables GC.
+- Any positive value enables GC and determines the number of epochs of historical data to retain.
+
+#### Recommendations
+
+1. **Archival Nodes**: **Keep GC disabled** (`GCRetentionEpochs` = 0) to retain all indexed data.
+
+2. **Non-Archival Nodes**:  Set `GCRetentionEpochs` to match the amount of chain state your node retains 
+(*Example:* if your node is configured to retain 2 days of Filecoin chain state with the Splitstore, set `GCRetentionEpochs` to `retentionDays * epochsPerDay = 2 * 2880 = 5760`).
+**Warning:** Setting this value below the chain state retention period will degrade RPC performance and reliability because the Index will lack data for epochs still present in the chain state.  
+
+### Removed Options
+
+**Note: The following config options no longer exist in Lotus and have been removed in favor of the ChainIndexer config options explained above. They can be removed when upgrading to Lotus v1.31.0.**
+
+```toml
+[Fevm]
+EthTxHashMappingLifetimeDays = 0
+
+[Events]
+DisableHistoricFilterAPI = false
+DatabasePath = ""
+```
+
+## Migration Guide
+
+Migrating to the new `ChainIndexer` involves several steps to ensure a smooth transition:
```

Comment:
Should the first step, especially for those who kept a large history of index, to back up their existing index sqlite db? I think it is important for archival indexer provider  to be able to rollback by (1) downgrade their lotus) (2) load the index db back upon node restart (3) have a functional node with existing system  without have to refill everything to minimize the service interruption due to indexer upgrade failuer


Now that i learnt that the migration and api switch is mandatory. For indexer providers,  I think it is particularly crucial for archival indexer providers instruction should start with `Back up the existing index SQLite database before updating your node to v1.3x.0`.

And we need to make the rollback strategy to be more robust as well..
- Downgrade Lotus to the previous version.
- Restore the backed-up index database upon node restart.
- Resume operations with the existing system, avoiding the need to refill all data.

This process is necessary to minimize service interruptions in case of unforeseen issues during the indexer upgrade.

Reply:
@jennijuju 

"avoiding the need to refill all data."

-> This will need us to build something like re-conciliation for all the existing Indices (there are three of them). Reconciliation is a complex process and investing the time to build it on top of a broken Indexing system is not something I agree with. This is a rare case that can be handled by the existing backfilling tooling we have.

Reply:
@jennijuju : I think there is some misunderstanding:

Lotus <1.31.0
Indexes live in LOTUS_HOME/sqllite

Lotus >=1.31.0 (which this PR will be part of)
Indexes live in LOTUS_HOME/chainindex

A user doesn't need to take a backup because we don't touch their old files.  Assuming they don't wipe away their old files, they can just downgrade to previous Lotus version and cary on.  No backup taking or restoring needed since we never touched the pre-1.31.0 data.  So I think we're good here procedurally.

I am happy to take fresh eyes later to look to see if we can communicate this better.  I can also jump into Fil Office to verbally sync.

Reply:
@BigLep i see & thanks for spotting the misunderstanding, i also noticed that it is mentioned in the later section "_It's recommended to keep the legacy directory around until you've confirmed you don't need to [rollback](#rollback)._"

It could be nice to call this out sooner and flag to operators they should make sure there is enough disk space for a separate db (if the db is small, then we can neglect this point).

and their should be another step added in this session "X confirms a successful migration, and you can remove the legacy index by doing Y once you get that confirmation

Reply:
(i think the key misunderstanding from my side the use of the word "migration", where in the past most index migration we have done was to migrate the existing DB to a new schema (and they were.. scary).  this seems like i fell into the 🕳️ of having too much historical context and I don't think we necessarily need to change anything here to communicate this better to the user..

Reply:
This is good feedback @jennijuju and makes sense.  I pushed some changes to hopefully make this clearer.  Can you please look at the file again (or the diff: https://github.com/filecoin-project/lotus/pull/12450/files/47e421f1f3e16c08f3b51a9f4da256a6876357d8..926852a538d50b40e8bdac552215aa3912526f90 )?